rfcs/47-relaxed-ordering-constraints.md

```diff
@@ -0,0 +1,172 @@
+
+## Drawbacks
+
+- Relaxed ordering constraints will be more expensive to initialize, increasing the cost of any schedule intiialization or modification.
+
+## Rationale and alternatives
+
+### Why is relaxed the correct default ordering strategy?
+
+Strict ordering is simple and explicit, and will never result in strange logic errors.
+On the other hand, it *will* result in pointless and surprising blocking behavior, possibly leading to unsatisfiable schedules.
+
+Relaxed ordering is the correct strategy in virtually all cases: in Bevy, interior mutability at the component or resource level is rare, almost never needed and results in other serious and subtle bugs.
+
+As we move towards specifying system ordering dependencies at scale, it is critical to avoid spuriously breaking users schedules, and silent, pointless performance hits are never good.
+
+### ## Should relaxed ordering constraints be transitive?
+
+If `A` is before `B`, and `B` is before `C`, must `A` be before `C`?
+
+Naturally, this seems like it must be the case: strict ordering constraints are transitive, and time is well-ordered after all!
+But if either the `A-B` or `B-C` edges are dissolved due to compatibility (and the `A-C` edge matters due to incompatibility), then `C` could be executed before `A`!
+
+There are two possible models that we could handle this important edge case:
+
+- **Model 1:** Infer relaxed-ordering constraints over each subgraph.
+  - Transitivity holds, as expected.
+  - Non-transitive relaxed ordering constraints cannot be represented.
+  - Some computational overhead, as we must create, evaluate and then deduplicate relaxed ordering constraints over each subgraph.
+- **Model 2:** Do nothing, report execution order ambiguities and allow the user to resolve this by adding additional constraints.
+  - Reduces risk of accidentally creating unnecessary constraints.
+  - More explicit, and thus more verbose.
+  - Relies on users checking and resolving execution order ambiguities.
+
+But which behavior is correct?
+Consider the following concrete case:
+
+We have five relevant systems:
+
+1. `determine_player_movement`: reads `Input`, writes `PlayerIntent`
+2. `collision_detection`: reads `Position` and `Velocity`, writes `Events<Collisions>`
+3. `collision_handling`: reads `Events<Collision>`, writes `Velocity`
+4. `apply_player_movement:` reads `PlayerIntent`, writes `Velocity`
+5. `apply_velocity:` reads `Velocity`, writes `Position`
+
+Suppose our user specifies the following relaxed ordering constraints:
+
+1. `determine_player_movement` is before `collision_detection`
+   1. Spurious: these systems are compatible.
+2. `collision_detection` is before `collision_handling`
+   1. Real: these systems conflict on `Events<Collisions>`
+3. `collision_handling` is before `apply_velocity`
+   1. Real: these systems conflict on `Velocity`
+4. `apply_player_movement` is after `collision_handling`
+   1. Spurious: these systems are compatible.

```

Comment:
Why are these systems compatible? Since you cannot reason about what `apply_player_movement` exactly does to `Velocity`, it should be treated as "reads and writes `Velocity`".
Imagine the following:
```rs
fn apply_player_movement(_: Res<&PlayerIntent>, velocity: Query<(&mut Velocity, With<Player>)>) {
  if velocity > 5 {
    // doesn't really matter what happens here,
    // it branches based on potentially stale data if this is considered spurious
  }
}
```